### DISCUSSION TITLE
Ist es möglich SOC bzw. PV Produnktions Daten per MQTT zu übertragen?

### DISCUSSION THREAD

**Discussion by MikeHF80** [hoylabs/OpenDTU-OnBattery#581] [Q&A]
Hallo zusammen,

mein Daten der PV Anlage (Victron) werden per Modbus in ioBroker übertragen. 
Gibt es eine Möglichkeit SOC und eventuell die gerade Produzierte PV Leistung per MQTT an openDTU on Battery zu übertragen und so die Produzierte Leistung des Wechselrichters zu begrenzen so das die Batterie nicht tagsüber entleert wird?

Viele Grüße
Mike

**Comment by Manos1966** [comment]
Hallo @MikeHF80 

ich kann auf deine Frage keine 100% prozentig qualifizierte Antwort geben, dafuer muess einer der Programmierer beantworten.
"SOC bzw. PV Produnktions Daten per MQTT zu übertragen?"
So fern ich weiss, nein!
Dafuer wurde OpenDTUonBattery nicht programmiert, sondern:

- VE.Direkt Daten per Kabel von Victron
- SOC per CANBus fuer Pylontech, oder SHUNT fuer Victron, oder Spannung fuer andere Batterien

Aber den Zweck deiner Frage verstehe ich nicht ganz!
Du willst: "so die Produzierte Leistung des Wechselrichters zu begrenzen so das die Batterie **nicht tagsüber entleert** wird?"

Dann einfach die Einstellung "Leeren zur Nach" auswählen unter Dynamic Power Limiter Settings. Ziel erreicht!
![Capture](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/689f7566-b363-4740-923a-a38dd162b98a)

Daraufhin wird deine Batterie tagsüber nicht benutzt.

**Reply by MikeHF80** [reply]
Hallo Manos,

ok mein Formulierung war eher nicht so gut.
Also angenommen die PV Module (Laderegler) produzieren 800Watt der Wechselrichter hat eine Anforderung von 1000Watt dann würden die fehlenden 200Watt aus der Batterie genommen.

Wenn ich das jetzt richtig versteh verhindert dies die Solar-Passthrough Option. Richtig?

Da ich aber ein Victron System mit 3 Stück MPPT einem Batterie Monitor (BMV-700) der mir die SOC; Spannung und Strom der Batterie liefert und eine Cerbo GX habe, habe ich das Problem das ich weder die MPPT´s noch den Batterie Monitor per Ve. Direkt an die OpenDTU on Battery anschließen kann.

Die Grid Zähler daten schicke ich per MQTT zu OpenDTU wenn ich jetzt noch die Laderegler Werte (Aktuelle PV Leistung) an die OpenDTU senden könnte dann könnte ich auch Solar-Passthrough nutzen.
Noch besser ich könnte auch noch den SOC; Batterie Spannung usw. per MQTT senden, dann wären alle Werte in der OpenDTU die benötigt werden um sauber zu Regeln.

Ich hoffe ich konnte mein Problem nun besser erklären.

Viele Grüße
Mike

**Reply by Manos1966** [reply]
Kommentar Entfernt da es nur bei "full solar-passthrough" zutrifft und nicht bei "(normal) solar-passthrough"

**Reply by schlimmchen** [reply]
solar passthrough ist dafür da, selbst verbrauchte energie direkt zu erzeugen, statt erst in die batterie zu laden, bis diese die startschwelle erreicht hat.

was du beschreibst ist "full solar passthrough", wofür es eine eigene schwelle gibt.

@MikeHF80 deine anforderung ist nachvollziehbar. SoC per MQTT liefern könnte ich vergleichsweise leicht umsetzen. aber beim PV ertrag wirds dann schwierig, weil es da noch keinerlei abstraktion gibt. da gibts VE.Direct oder nichts.

**Reply by Manos1966** [reply]
Danke @schlimmchen , ich war voll auf dem falschen Pferd, Mist 🤦‍♂️

**Reply by MikeHF80** [reply]
Den SOC per MQTT an OpenDTU zu senden wäre schon mal sehr hilfreich. 

Was es nun bedeutet das der PV Ertrag nur per VE.Direkt verarbeitet wird verstehe ich ehrlich gesagt nicht.
Habe gedacht das der Wert einfach eine Variable ist die man theoretisch von verschiedener Stelle aus setzen könnte.

Viele Grüße 
Michael

**Reply by schlimmchen** [reply]
@MikeHF80 Siehe #589.

> Habe gedacht das der Wert einfach eine Variable ist die man theoretisch von verschiedener Stelle aus setzen könnte.

Mir ist schleierhaft, wie du zu dieser Einschätzung kommst, jedenfalls ist sie zu simpel. Bei der von mir angesprochenen Abstraktion geht es insb. in der GUI darum, dass dem Nutzer klar wird, wie da was eingestellt wird. Bei der Batterie ist das inzwischen meiner Meinung nach sehr gut umgesetzt: Man kann die Schnittstelle zur Batterie einschalten, und dann wählt man aus, wie mit der Batterie gesprochen wird. Bei VE.Direct heißt der Menüpunkt eben "VE.Direct". Und auch intern heißt alles "VeDirect" oder ähnlich. Dabei will man nun plötzlich, dass VeDirect nur eine spezielle Ausprägung von etwas Abstrakterem ist. Und auch das Abstrakte muss erst mal einen vernünftigen Namen erhalten. Und zwar einen, der beim Hinzufügen der nächsten Technologie nicht wieder veraltet ist. "Charge Controller" passt zum Beispiel nicht. Das ist zwar etwas abstrakter, aber "Solare Leistung aus dem MQTT Broker" ist nicht ein "Charge Controller". Ein Menüpunkt "Solar Power" ist auch doof, wer vermutet denn dort die Einstellungen für den per VE.Direct angeschlossenen Charge Controller? Also wenn du hier einen guten Vorschlag hast wäre das eine großartige Unterstützung.

**Reply by sequarell** [reply]
Hallo zusammen erstmal.. Ich war hier in dem Projekt noch nciht aktiv, aber möchte mal in der Diskussion mit einsteigen. Ich habe selbst zuhause openDTU onBattery im Einsatz und ich habe für mein Setup im Grunde  die selbe Anforderung von @MikeHF80 umgesetzt. Ich habe uach mehrere Laderegler (vier gesamt) und habe die Gesamtleistung auf einem MQTT Topic. 
Was @schlimmchen gesagt hat mag im ersten Moment etwas ruppig klingen, ist aber aus SW-Entwicklersicht erstmal richtig. Ich selber habe für meine eigene Installation die Firmware modifiziert, indem ich die VEDirect-Klasse ienfach "ausgehöhlt" habe und dort die Daten von MQTT hole.. ist aber wirklich in der Form für einen Endnutzer absolut **nicht** zumutbar/brauchbar.
Trotzdem habe ich halt schon einen gewissen Teil der Probleme gelöst und würde das bei Interesse gerne in das Projekt zurückfließen lassen. Meine Zeit ist leider momentan zwar sehr begrenzt, aber ich würde dem Projekt gerne was zurückgeben :)
@schlimmchen Wie würdest Du es sehen: Wenn man die Abstraktion z.B. DC-Input nennt? Dahinter könnten sich dann verschiedenste Laderegler verstecken, egal ob gespeist von Solar, Wind, Netz..

**Reply by schlimmchen** [reply]
Diese Abstraktion wird es geben, keine Frage. DC-Input ist schwierig, weil der Huawei AC Lader ja im Grunde darunter fallen würde. Ich denke es geht spezieller im "Solarladeregler". Aber jetzt nennst du Wind... Die Benamung ist durchaus knifflig. Nur "Laderegler"?

Natürlich darfst du gerne deine Arbeit präsentieren und einen Pull Request erstellen, dann schauen wir uns das an.

Wenn du deine Victron SmartSolar Laderegler in ein VE.Smart Netzwerk steckst, und einen davon mit der OpenDTU-OnBattery verbindest (sowohl RX als auch TX), dann wird es ab dem nächstes Release die Funktion geben, dass die Netzwerk-Gesamtleistung abgefragt und verwendet wird.

Der Wunsch, sie solare Leistung auch per MQTT abzurufen, ist dann trotzdem noch anzugehen. Das ist auch nicht kompliziert, ich hab nur keine Lust drauf, weil erst die Abstraktion eingeführt werden muss, die entsprechend Zeit braucht.

**Reply by sequarell** [reply]
> ich hab nur keine Lust drauf, weil erst die Abstraktion eingeführt werden muss, die entsprechend Zeit braucht.

Kann ich sehr gut nachvollziehen. Ich hab es wie gesagt für mich auch nur "hingehackt". Im Grunde habe ich eine Anbindung an VenusOS realisiert, dort liegen ja alle relevanten Daten.
Wenn ich beruflich mal Luft habe, werde ich mich hinsetzen und die Fleißarbeit bzgl. der Abstraktionsebene auf mich nehmen..

**Comment by Manos1966** [comment]
As a side point: You have the Data in IO Broker.

You can control the OPen DTU Power Limiter by sending MQTT commands to it.
Look here:
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/MQTT#power-limiter-topics

**Reply by MikeHF80** [reply]
Thanks for the info, but unfortunately this doesn't solve my problem. Since the regulation should remain with OpenDTU on Battery, I only want to send SOC, PV power, etc. to OpenDTU via MQTT. because I feel it is possible to connect the charge controllers and SmartShunt directly to OpenDTU.

**Comment by schlimmchen** [comment]
@MikeHF80 Did you notice #589?
